Add initial changes to cep-47 introduction and prepare sections (#196)
casper-network/docs: source/docs/casper/dapp-dev-guide/tutorials/cep47/index.md

> # Introduction

Reviewer: ```suggestion
# The NFT Standard on Casper (CEP-47)
```

> This tutorial takes you through the NFT implementation of the Casper Network. Before diving deep into the Casper NFT implementation, let's cover some knowledge on overall NFT design. The following sections illustrate the origin and the development of NFTs in the blockchain domain.

Reviewer: ```suggestion
This tutorial takes you through the standard of implementing [non-fungible tokens](/docs/glossary/N#non-fungible-token) on the Casper Network. Before diving into the Casper NFT implementation, let's cover some knowledge on overall NFT design. The following sections illustrate the origin and the development of NFTs in the blockchain domain.
```

Reviewer: I moved the NFT definition up, where it is first used.

> ### ERC-721 Standard

Reviewer: ```suggestion
### The NFT Standard on Ethereum
```

> ERC-721 is the standard defined to implement the [non-fungible tokens](/docs/glossary/N#non-fungible-token) inside Ethereum blockchain. This standard defines a minimum interface a smart contract should implement to manage, own and trade the unique tokens.  

Reviewer: ```suggestion
[ERC-721](https://eips.ethereum.org/EIPS/eip-721) is the standard defined to implement NFTs on the Ethereum blockchain. This standard defines a minimum interface a smart contract should implement to manage, own, and trade unique tokens. 
```

Reviewer: Please double check that links and definitions are mentioned when a term is first used (such as ERC-721, CEP-47).

> ## ERC-721 Implementation within Casper Network

Reviewer: ```suggestion
## The NFT Standard on Casper
```

> While [ERC-721](https://eips.ethereum.org/EIPS/eip-721) is the original name for the standard given in Ethereum blockchain, CEP-47 is the customized standard for the Casper platform. CEP-47 standard supports the NFT behaviors inside the Casper network. Casper ecosystem needs an NFT standard to enable further development of the NFT-based products. Due to the unique features within the Casper VM, the standards originate from Ethereum ecosystem are not the best choice for the Casper platform. Hence, CEP-47 is built to support these deviations and it is mainly focused on a gas-efficient design to support those unique traits.

Author: > `the standards originate from Ethereum ecosystem are not the best choice for the Casper platform` is horrible phrasing, suggests that we cannot do what eth can

Thanks, @deuszex - I grabbed that from PR 47 (file - [./0047-casper-nft-protocol.md](https://github.com/casper-network/ceps/pull/47/files#) -> line 12-14 ). Could you please suggest a good alternative way to describe it or do you think it's better to remove that part?

> While [ERC-721](https://eips.ethereum.org/EIPS/eip-721) is the original name for the standard given in Ethereum blockchain, CEP-47 is the customized standard for the Casper platform. CEP-47 standard supports the NFT behaviors inside the Casper network. Casper ecosystem needs an NFT standard to enable further development of the NFT-based products. Due to the unique features within the Casper VM, the standards originate from Ethereum ecosystem are not the best choice for the Casper platform. Hence, CEP-47 is built to support these deviations and it is mainly focused on a gas-efficient design to support those unique traits.

Reviewer: ```suggestion
[CEP-47](https://github.com/casper-ecosystem/casper-nft-cep47) is the NFT standard for the Casper blockchain, supporting the unique traits of the Casper Virtual Machine and its gas-efficient design. 
```

> While [ERC-721](https://eips.ethereum.org/EIPS/eip-721) is the original name for the standard given in Ethereum blockchain, CEP-47 is the customized standard for the Casper platform. CEP-47 standard supports the NFT behaviors inside the Casper network. Casper ecosystem needs an NFT standard to enable further development of the NFT-based products. Due to the unique features within the Casper VM, the standards originate from Ethereum ecosystem are not the best choice for the Casper platform. Hence, CEP-47 is built to support these deviations and it is mainly focused on a gas-efficient design to support those unique traits.

Reviewer: I recommend simplifying this and keeping it short and sweet -- unless you can get more information from developers about the "unique traits" and describe those further. Proposed commit below.

> ## CEP-47 Standard

Reviewer: ```suggestion
```

Reviewer: Suggesting to remove the title and merge the content with the previous paragraph.

> CEP47 takes full advantage of the URef feature to optimize smart contracts interactions. It also uses the Groups feature to optimize gas costs. Since this is a different type of token there are special endpoint implementations to support that unique features. You can find the [main functions](./index#cep-47-function-types) when going through the [contract file](https://github.com/casper-ecosystem/casper-nft-cep47/blob/master/cep47/bin/cep47_token.rs) inside the [CEP-47 repository](https://github.com/casper-ecosystem/casper-nft-cep47).

Author: This part is also taken from [Cep-47 - Casper NFT Protocol](https://github.com/casper-network/ceps/pull/47/files#) file , line numbers 16 17. Appreciate if you can re-phrase it in a better way, suggest any new idea or verify the deletion of that part. Thanks

> CEP47 takes full advantage of the URef feature to optimize smart contracts interactions. It also uses the Groups feature to optimize gas costs. Since this is a different type of token there are special endpoint implementations to support that unique features. You can find the [main functions](./index#cep-47-function-types) when going through the [contract file](https://github.com/casper-ecosystem/casper-nft-cep47/blob/master/cep47/bin/cep47_token.rs) inside the [CEP-47 repository](https://github.com/casper-ecosystem/casper-nft-cep47).

Reviewer: ```suggestion
CEP-47 takes full advantage of [unforgeable references](https://casper.network/docs/design/uref) to store values and manage permissions to them. The standard also takes advantage of other access control features (such as groups). We recommend exploring the [main functions](./index#cep-47-function-types) of the [contract](https://github.com/casper-ecosystem/casper-nft-cep47/blob/master/cep47/bin/cep47_token.rs) to understand the standard further.
```

> CEP47 takes full advantage of the URef feature to optimize smart contracts interactions. It also uses the Groups feature to optimize gas costs. Since this is a different type of token there are special endpoint implementations to support that unique features. You can find the [main functions](./index#cep-47-function-types) when going through the [contract file](https://github.com/casper-ecosystem/casper-nft-cep47/blob/master/cep47/bin/cep47_token.rs) inside the [CEP-47 repository](https://github.com/casper-ecosystem/casper-nft-cep47).

Reviewer: 1. I suggest simplifying (see commit suggestion), unless you can find out more about how those optimizations can be made. However, since this is an introduction, you can talk more about the optimizations in another section.
2. Please add the Groups explanation to the glossary, or somewhere in the design section: 
"The Groups feature provides access control to a contract's runtime context. It is used only once to restrict the use of the constructor entry_point, which sets up necessary data storage in the contract context."
3. Once you have a link, please add it as a reference where you first mention "Groups".

### CEP-47 Function Types

> CEP-47 standard contains endpoints to enable the NFT functionalities. The basic attributes of tokens are addressed by *name*, *symbol*, *total_supply*, and *balance_of* endpoints.The token ownership is managed by *owner_of*, *get_token_by_index* endpoints. Token metadata is handled  through *meta*, *token_meta*, *update_token_meta* endpoints. Main behaviors of NFTs are handled through *mint*, *mint_copies*, *burn*, *transfer*, *transfer_from*, *approve*, *get_approved* endpoints.

Reviewer 2: `name`, `symbol`, `meta`, `total_supply` are details of the whole contract.
`balance_of`, `get_token_by_index` get details of tokens related to a specific account.
`owner_of`, `token_meta` are for details of a specific token.
All of the above can only be called from inside contracts, as they return data.

Reviewer: @Buddhika-Pathirathna I recommend changing this paragraph by adding a list and a short sentence about each function. Then add what @deuszex mentioned as a note.

```
The CEP-47 standard contains the following endpoints:
*name* - ...
*symbol* - ...
*total_supply* - ...
*balance_of* - ...
*owner_of* - ...
*get_token_by_index* - ...
*meta* - ...
*token_meta* - ...
*update_token_meta*  - ...
*mint* - ...
*mint_copies* - ...
*burn* - ...
*transfer* - ...
*transfer_from* - ...
*approve* - ...
*get_approved*  - ...

:::note
These functions can only be called from inside contracts, as they return data:
*name*, *symbol*, *meta*, and *total_supply* return details regarding the whole contract
*balance_of* and *get_token_by_index* retrieve details of tokens related to a specific account
*owner_of* and *token_meta* retrieve the details of a specific token
:::
```

> If you haven't read [Writing Rust Contracts on Casper](../../writing-contracts/rust.md), we recommend you start there to learn more about smart contracts.

Reviewer: ```suggestion
```

Reviewer: I think it is best to remove this line.

# The NFT Standard on Casper (CEP-47)

This tutorial takes you through the standard of implementing [non-fungible tokens](/docs/glossary/N#non-fungible-token) on the Casper Network. Before diving into the Casper NFT implementation, let's cover some knowledge on overall NFT design. The following sections illustrate the origin and the development of NFTs in the blockchain domain.

### The NFT Standard on Ethereum
[ERC-721](https://eips.ethereum.org/EIPS/eip-721) is the standard defined to implement NFTs on the Ethereum blockchain. This standard defines a minimum interface a smart contract should implement to manage, own, and trade unique tokens. 

## The NFT Standard on Casper
[CEP-47](https://github.com/casper-ecosystem/casper-nft-cep47) is the NFT standard for the Casper blockchain, supporting the unique traits of the Casper Virtual Machine and its gas-efficient design. 

> CEP-47 takes full advantage of [unforgeable references](/docs/design/uref) to store values and manage permissions to them. The standard also takes advantage of other access control features (such as [groups](/docs/glossary/G#groups)). We recommend exploring the [main functions](../cep47/#cep-47-function-types) of the [contract](https://github.com/casper-ecosystem/casper-nft-cep47/blob/master/cep47/bin/cep47_token.rs) to understand the standard further.

Reviewer: ```suggestion
The Casper NFT standard takes full advantage of [unforgeable references](/docs/design/uref) to store values and manage permissions to them. It also takes advantage of other access control features (such as [groups](/docs/glossary/G#groups)). We recommend exploring the [main functions](../cep47/#cep-47-function-types) of the [contract](https://github.com/casper-ecosystem/casper-nft-cep47/blob/master/cep47/bin/cep47_token.rs) to understand the standard further.
```

> ### CEP-47 Function Types

Reviewer: ```suggestion
### CEP-47 Functions
```

> CEP-47 standard contains following endpoints to enable the NFT functionalities. 

Reviewer: ```suggestion
The CEP-47 standard contains the following functions to enable NFTs. 
```

> The CEP-47 standard contains the following endpoints:

Reviewer: Delete this line, as it seems to be duplicated.
```suggestion
```

> Metadata endpoints will be used by other smart contracts

Reviewer: I think this line should also be deleted.
```suggestion
```

- [*name*](https://github.com/casper-ecosystem/casper-nft-cep47/blob/09b40b0caf4cfc6f73d1e5f7d5b9c868228f7621/cep47/bin/cep47_token.rs#L44-L47) - Returns the name of the NFT contract
- [*symbol*](https://github.com/casper-ecosystem/casper-nft-cep47/blob/09b40b0caf4cfc6f73d1e5f7d5b9c868228f7621/cep47/bin/cep47_token.rs#L50-L53) - Returns the symbol of the NFT contract

> - [*meta*](https://github.com/casper-ecosystem/casper-nft-cep47/blob/09b40b0caf4cfc6f73d1e5f7d5b9c868228f7621/cep47/bin/cep47_token.rs#L56-L59) - Returns the metadata of NFT

Reviewer: ```suggestion
- [*meta*](https://github.com/casper-ecosystem/casper-nft-cep47/blob/09b40b0caf4cfc6f73d1e5f7d5b9c868228f7621/cep47/bin/cep47_token.rs#L56-L59) - Returns the metadata of the NFT contract
```

- [*total_supply*](https://github.com/casper-ecosystem/casper-nft-cep47/blob/09b40b0caf4cfc6f73d1e5f7d5b9c868228f7621/cep47/bin/cep47_token.rs#L62-L65)- Returns the amount of issued NFTs

> - [*balance_of*](https://github.com/casper-ecosystem/casper-nft-cep47/blob/09b40b0caf4cfc6f73d1e5f7d5b9c868228f7621/cep47/bin/cep47_token.rs#L68-L72) - Returns an amount of NFT instances the `owner` owns

Reviewer: ```suggestion
- [*balance_of*](https://github.com/casper-ecosystem/casper-nft-cep47/blob/09b40b0caf4cfc6f73d1e5f7d5b9c868228f7621/cep47/bin/cep47_token.rs#L68-L72) - Returns the amount of NFT tokens the `owner` holds
```

> - [*get_token_by_index*](https://github.com/casper-ecosystem/casper-nft-cep47/blob/09b40b0caf4cfc6f73d1e5f7d5b9c868228f7621/cep47/bin/cep47_token.rs#L75-L80) - Getter for token index. Tokens are stored by tokenId and by an index

Reviewer: I don't understand this explanation. Do you mean this:

```suggestion
- [*get_token_by_index*](https://github.com/casper-ecosystem/casper-nft-cep47/blob/09b40b0caf4cfc6f73d1e5f7d5b9c868228f7621/cep47/bin/cep47_token.rs#L75-L80) - Retrieves the NFT token at a specific index
```

> - [*owner_of*](https://github.com/casper-ecosystem/casper-nft-cep47/blob/09b40b0caf4cfc6f73d1e5f7d5b9c868228f7621/cep47/bin/cep47_token.rs#L83-L87) - Getter for the owner of a given token

Reviewer: ```suggestion
- [*owner_of*](https://github.com/casper-ecosystem/casper-nft-cep47/blob/09b40b0caf4cfc6f73d1e5f7d5b9c868228f7621/cep47/bin/cep47_token.rs#L83-L87) - Retrieves the owner of a given token
```

> - [*token_meta*](https://github.com/casper-ecosystem/casper-nft-cep47/blob/09b40b0caf4cfc6f73d1e5f7d5b9c868228f7621/cep47/bin/cep47_token.rs#L90-L94) - Getter for the metadata for a given token

Reviewer: ```suggestion
- [*token_meta*](https://github.com/casper-ecosystem/casper-nft-cep47/blob/09b40b0caf4cfc6f73d1e5f7d5b9c868228f7621/cep47/bin/cep47_token.rs#L90-L94) - Retrieves the metadata for a given token
```

- [*update_token_meta*](https://github.com/casper-ecosystem/casper-nft-cep47/blob/09b40b0caf4cfc6f73d1e5f7d5b9c868228f7621/cep47/bin/cep47_token.rs#L97-L103) -  A function to update the metadata of a token

> - [*mint*](https://github.com/casper-ecosystem/casper-nft-cep47/blob/09b40b0caf4cfc6f73d1e5f7d5b9c868228f7621/cep47/bin/cep47_token.rs#L106-L113) - Creates a list of tokens with the given ids and the given metas for one given recipient. tokenIds and metas are paired in order

Reviewer: ```suggestion
- [*mint*](https://github.com/casper-ecosystem/casper-nft-cep47/blob/09b40b0caf4cfc6f73d1e5f7d5b9c868228f7621/cep47/bin/cep47_token.rs#L106-L113) - Creates a list of tokens for a specific recipient, given the token IDs and their metadata, paired in order
```

> - [*mint_copies*](https://github.com/casper-ecosystem/casper-nft-cep47/blob/09b40b0caf4cfc6f73d1e5f7d5b9c868228f7621/cep47/bin/cep47_token.rs#L116-L124) - Creates for an address, several tokens with the given tokenIds but with the same meta

Reviewer: ```suggestion
- [*mint_copies*](https://github.com/casper-ecosystem/casper-nft-cep47/blob/09b40b0caf4cfc6f73d1e5f7d5b9c868228f7621/cep47/bin/cep47_token.rs#L116-L124) - For a given address, this function creates several tokens with specific IDs but with the same metadata
```

> - [*burn*](https://github.com/casper-ecosystem/casper-nft-cep47/blob/09b40b0caf4cfc6f73d1e5f7d5b9c868228f7621/cep47/bin/cep47_token.rs#L127-L133) - Destroys the given tokens for a given account

Reviewer: ```suggestion
- [*burn*](https://github.com/casper-ecosystem/casper-nft-cep47/blob/09b40b0caf4cfc6f73d1e5f7d5b9c868228f7621/cep47/bin/cep47_token.rs#L127-L133) - Destroys the tokens in the account given
```

- [*transfer*](https://github.com/casper-ecosystem/casper-nft-cep47/blob/09b40b0caf4cfc6f73d1e5f7d5b9c868228f7621/cep47/bin/cep47_token.rs#L136-L142) - Transfers tokens to another account
- [*transfer_from*](https://github.com/casper-ecosystem/casper-nft-cep47/blob/09b40b0caf4cfc6f73d1e5f7d5b9c868228f7621/cep47/bin/cep47_token.rs#L145-L152) - Transfer tokens from a given account to another account

> - [*approve*](https://github.com/casper-ecosystem/casper-nft-cep47/blob/09b40b0caf4cfc6f73d1e5f7d5b9c868228f7621/cep47/bin/cep47_token.rs#L155-L161) - Give an account 'approved rights' to transfer tokens that you hold

Reviewer: Suggestion to write this in plain English:
```suggestion
- [*approve*](https://github.com/casper-ecosystem/casper-nft-cep47/blob/09b40b0caf4cfc6f73d1e5f7d5b9c868228f7621/cep47/bin/cep47_token.rs#L155-L161) - Gives another account the right to transfer tokens from this account
```

> - [*get_approved*](https://github.com/casper-ecosystem/casper-nft-cep47/blob/09b40b0caf4cfc6f73d1e5f7d5b9c868228f7621/cep47/bin/cep47_token.rs#L164-L169) - Getter function to see if the owner has given you approved rights to the required tokens 

Reviewer: Suggestion to write this in plain English:
```suggestion
- [*get_approved*](https://github.com/casper-ecosystem/casper-nft-cep47/blob/09b40b0caf4cfc6f73d1e5f7d5b9c868228f7621/cep47/bin/cep47_token.rs#L164-L169) - Retrieves information about the rights to transfer tokens from another account
```
